### ISSUE TITLE
[JENKINS-45138] How do I `lock` outside of a step in a declarative pipeline job?

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/lockable-resources-plugin#886]
<p>In the freestyle/UI-based job configuration, it looks like a lock is scoped to the entire build, not one step.</p>

<p> <span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38649/38649_freestyle-configuration.png" style="border: 0px solid black" /></span></p>

<p>The <a href="https://wiki.jenkins.io/display/JENKINS/Lockable+Resources+Plugin" class="external-link" target="_blank" rel="nofollow noopener">documentation</a> for pipelines seems to imply that `lock` must be used within a `step`, but could you confirm?</p>

<p>I have a situation where I need to lock a resource while performing a step. I need to always cleanup that resource, so that my step works correctly.</p>

<pre>
step(<span class="code-quote">'foo'</span>) {
    lock(<span class="code-quote">'foo-resource'</span>) {
bat <span class="code-quote">'... acquire the resource...'</span>
bat <span class="code-quote">'... use the resource...'</span>
bat <span class="code-quote">'... cleanup the resource...'</span>
    }
}
</pre>

<p>I need to move the cleanup step to a `post / always`, so the step reports the correct exit code, but now it's outside of the lock.</p>

<pre>
step(<span class="code-quote">'foo'</span>) {
    lock(<span class="code-quote">'foo-resource'</span>) {
bat <span class="code-quote">'... acquire the resource...'</span>
bat <span class="code-quote">'... use the resource...'</span>
    }
}
post {
    always {
bat <span class="code-quote">'... cleanup the resource...'</span>
    }
}
</pre>

<p>I've got other jobs waiting for this resource and, based on the timing, they're trying to use the resource that this job is now cleaning up.</p>

<hr />

<p>Well, I guess it's not valid around a `stage`.</p>

<pre>
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: 30: Unknown stage section <span class="code-quote">"lock"</span>. Starting with version 0.5, steps in a stage must be in a steps block. @ line 30, column 9.
   stage(<span class="code-quote">'foo'</span>) {
   ^

WorkflowScript: 30: Nothing to execute within stage <span class="code-quote">"foo"</span> @ line 30, column 9.
   stage(<span class="code-quote">'foo'</span>) {
   ^
</pre>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=anthonymastrean">anthonymastrean</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45138" target="_blank">How do I `lock` outside of a step in a declarative pipeline job?</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: lockable-resources-plugin
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2017-06-29T15:46:13+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 20251217-081411
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In the freestyle/UI-based job configuration, it looks like a lock is scoped to the entire build, not one step.</p>

<p> <span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38649/38649_freestyle-configuration.png" style="border: 0px solid black" /></span></p>

<p>The <a href="https://wiki.jenkins.io/display/JENKINS/Lockable+Resources+Plugin" class="external-link" target="_blank" rel="nofollow noopener">documentation</a> for pipelines seems to imply that `lock` must be used within a `step`, but could you confirm?</p>

<p>I have a situation where I need to lock a resource while performing a step. I need to always cleanup that resource, so that my step works correctly.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">step(<span class="code-quote">'foo'</span>) {
    lock(<span class="code-quote">'foo-resource'</span>) {
        bat <span class="code-quote">'... acquire the resource...'</span>
        bat <span class="code-quote">'... use the resource...'</span>
        bat <span class="code-quote">'... cleanup the resource...'</span>
    }
}
</pre>
</div></div>

<p>I need to move the cleanup step to a `post / always`, so the step reports the correct exit code, but now it's outside of the lock.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">step(<span class="code-quote">'foo'</span>) {
    lock(<span class="code-quote">'foo-resource'</span>) {
        bat <span class="code-quote">'... acquire the resource...'</span>
        bat <span class="code-quote">'... use the resource...'</span>
    }
}
post {
    always {
        bat <span class="code-quote">'... cleanup the resource...'</span>
    }
}
</pre>
</div></div>

<p>I've got other jobs waiting for this resource and, based on the timing, they're trying to use the resource that this job is now cleaning up.</p>

<hr />

<p>Well, I guess it's not valid around a `stage`.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: 30: Unknown stage section <span class="code-quote">"lock"</span>. Starting with version 0.5, steps in a stage must be in a steps block. @ line 30, column 9.
           stage(<span class="code-quote">'foo'</span>) {
           ^

WorkflowScript: 30: Nothing to execute within stage <span class="code-quote">"foo"</span> @ line 30, column 9.
           stage(<span class="code-quote">'foo'</span>) {
           ^
</pre>
</div></div></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins ver. 2.63<br/>
Lockable Resource Plugin v2.0
```
</details>
<details><summary><i>1 attachment</i></summary>

- [freestyle-configuration.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38649/freestyle-configuration.png)
  > ![freestyle-configuration.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38649/freestyle-configuration.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vas:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45138?focusedId=361003&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-361003">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a>, this issue was marked as duplicate of 43336, but actually is not.</p>

<p>The current issue says that the lock does not include the post section.</p>

<p>In our case I have something like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
stage(<span class="code-quote">'tests'</span>) {
  options {
    lock(<span class="code-quote">'something'</span>)
  }
  
  parallel {
    stage(<span class="code-quote">'run1'</span>) {...}
    stage(<span class="code-quote">'run2'</span>) {...}
  }
  post {
    cleanup {...}
  }
} </pre>
</div></div>
<p>However, the port runs after releasing the lock (so, I get some errors when two build run in parallel).</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a>, this issue was marked as duplicate of 43336, but actually is not.</p>

<p>The current issue says that the lock does not include the post section.</p>

<p>In our case I have something like this:</p>

<pre>

stage(<span class="code-quote">'tests'</span>) {
  options {
    lock(<span class="code-quote">'something'</span>)
  }
  
  parallel {
    stage(<span class="code-quote">'run1'</span>) {...}
    stage(<span class="code-quote">'run2'</span>) {...}
  }
  post {
    cleanup {...}
  }
} </pre>
<p>However, the port runs after releasing the lock (so, I get some errors when two build run in parallel).</p>
